### ISSUE TITLE
`OpLoad Pointer <id> is not a logical pointer` when using PushConstant.

### ISSUE THREAD

**Issue by XAMPPRocky** [EmbarkStudios/rust-gpu#312]
```rust
use glam::{Mat4, Vec2, Vec4};
use spirv_std::{Image2d, Sampler, storage_class::{Input, Output, PushConstant, Uniform}};

#[allow(unused_attributes)]
#[spirv(fragment)]
pub fn main_fs(
    #[spirv(descriptor_set = 0, binding = 0)]  font_tex: Uniform<Image2d>,
    #[spirv(descriptor_set = 0, binding = 1)]  font_sampler: Uniform<Sampler>,
    in_uv: Input<Vec2>,
    in_color: Input<Vec4>,
    mut out_color: Output<Vec4>,
) {
    out_color.store(
        in_color.load() *
            font_tex.load()
                .sample(font_sampler.load(), in_uv.load())
    );
}

#[allow(unused_attributes)]
#[derive(Copy, Clone)]
#[repr(C)]
#[spirv(block)]
pub struct Constants {
    projection_matrix: Mat4,
}

#[allow(unused_attributes)]
#[derive(Copy, Clone)]
#[repr(C)]
#[spirv(block)]
pub struct GlPerVertex {
    gl_position: Vec4,
}

#[allow(unused_attributes)]
#[spirv(vertex)]
pub fn main_vs(
    in_pos: Input<Vec2>,
    in_uv: Input<Vec2>,
    in_color: Input<Vec4>,
    push_constants: PushConstant<Constants>,
    mut out_uv: Output<Vec2>,
    mut out_color: Output<Vec4>,
    mut gl_per_vertex: Output<GlPerVertex>,
) {

    out_uv.store(in_uv.load());
    out_color.store(linear_from_srgba(in_color.load()));

    gl_per_vertex.then(move |mut gl_per_vertex| {
        let in_pos = in_pos.load();
        gl_per_vertex.gl_position = 
            push_constants.load().projection_matrix *
            Vec4::new(0.0, 0.0, 0.0, 1.0);
            Vec4::new(in_pos.x, in_pos.y, 0.0, 1.0);
        // TODO: Change to use negative viewport height instead
        gl_per_vertex.gl_position.y = -gl_per_vertex.gl_position.y;
        gl_per_vertex
    });
}

pub fn linear_from_srgb(srgb: Vec3) -> Vec3 {
    let a = srgb / 12.92;
    let b = ((srgb + Vec3::splat(0.055)) / 1.055).powf(2.4);
    let c = {
        let x = Vec3::splat(0.04045);
        x.step(srgb)
    };

    mix(a, b, c)
}

pub fn linear_from_srgba(srgba: Vec4) -> Vec4 {
    let l = linear_from_srgb(srgba.into());
    Vec4::new(l.x, l.y, l.z, srgba.w)
}
```

### Error
```
error: internal error:0:0 - OpLoad Pointer <id> '1061[%1061]' is not a logical pointer.
  |
  = note: spirv-val failed
  = note: module "/Users/erin.power/src/rust_shaders.spv"

error: aborting due to previous error; 7 warnings emitted

error: could not compile `rust-shaders`
```

**Comment by eddyb** [comment]
Minimal repro:
```rust
use spirv_std::storage_class::{Input, Output};

#[allow(unused_attributes)]
#[spirv(vertex)]
pub fn main_vs(a: Input<f32>, b: Input<f32>, mut out: Output<f32>) {
    out.store((move || a.load() + b.load())());
}
```

Alternatively, since closures are just sugar:
```rust
#[allow(unused_attributes)]
#[spirv(vertex)]
pub fn main_vs(a: Input<f32>, b: Input<f32>, mut out: Output<f32>) {
    let tuple = (a, b);
    out.store(tuple.0.load() + tuple.1.load());
}
```

This works fine when the closure captures/`tuple` contains only one `Input`, but for two or more we get errors like this:
```
  error: invalid id:0:0 - OpLoad Pointer <id> '22[%22]' is not a logical pointer.
    %23 = OpLoad %float %22
```

So I'd say placing any of the `spirv_std::storage_class` types into an aggregate (including capturing it in a closure) is unsupported for now, but I'll look at the generated SPIR-V next, maybe this isn't a fundamental limitation.

**Comment by eddyb** [comment]
According to @khyperia on Discord, depending on where/whether references are taken, this might be fixed by SROA, or not even that. My guess is #300 would help with this, and require less specialized handling to make it work.

So for now the best thing to do is to load all the inputs and store all the outputs in `main_vs`.

<hr/>

I was going to suggest changing the code in the OP to this, but this is also broken:
```rust
    let in_pos = in_pos.load();
    let constants = push_constants.load();
    gl_per_vertex.then(|mut gl_per_vertex| {
        gl_per_vertex.gl_position = constants.projection_matrix * Vec4::new(0.0, 0.0, 0.0, 1.0);
        Vec4::new(in_pos.x, in_pos.y, 0.0, 1.0);
        // TODO: Change to use negative viewport height instead
        gl_per_vertex.gl_position.y = -gl_per_vertex.gl_position.y;
        gl_per_vertex
    });
```
```
  error: invalid id:0:0 - Pointer operand 260[%260] must be a memory object declaration
    %264 = OpFunctionCall %void %simplest_shader__main_vs___closure_0_ %252 %260 %261 %263
```

At least this works tho, so if the `then` helper causes issues because of the closure, you probably need to avoid it:
```rust
    let in_pos = in_pos.load();
    let mut gl_position = push_constants.load().projection_matrix * Vec4::new(0.0, 0.0, 0.0, 1.0);
    Vec4::new(in_pos.x, in_pos.y, 0.0, 1.0);
    // TODO: Change to use negative viewport height instead
    gl_position.y = -gl_position.y;
    gl_per_vertex.store(GlPerVertex { gl_position });
```

**Comment by khyperia** [comment]
Fixed by #690 :tada:
